stdlib/public/runtime/HeapObject.cpp

```diff
@@ -61,9 +61,11 @@ using namespace swift;
 /// Returns true if the pointer passed to a native retain or release is valid.
 /// If false, the operation should immediately return.
 static inline bool isValidPointerForNativeRetain(const void *p) {
-#if defined(__x86_64__) || defined(__arm64__) || defined(__aarch64__) || defined(_M_ARM64)
-  // On these platforms, the upper half of address space is reserved for the
+#if defined(__x86_64__) || defined(__arm64__) || defined(__aarch64__) || defined(_M_ARM64) || defined(__s390x__)
+  // On these platforms, except s390x, the upper half of address space is reserved for the
   // kernel, so we can assume that pointer values in this range are invalid.
+  // On s390x it is theoretically possible to have high bit set but in practice
```

Comment:
This last commit is a bit nerve-wracking. Is there somewhere we could add an assertion for s390x so if this ever goes awry, it's easy to debug?

Reply:
@jckarter and @mikeash convinced me that this should be okay, so I'll worry less.

Reply:
Thanks all for helping us narrow the s390x gap!

Reply:
Should be fine as long as swift_*alloc* never use the high half of address space. The issue here is whether a pointer to a refcountable Swift object is ever in the high half—if that's never the case, then it's OK.

Reply:
Is there a way to make this comment more descriptive then? Maybe mention that while the address space theoretically includes it, swift_alloc can never allocated objects at those addresses.

Reply:
@milseman I cannot be sure if swift_alloc can never allocate objects at those addresses.  Is there somewhere I can look to confirm?

Reply:
By default the allocator functions just call down to the platform `malloc`, so you could look at whatever `malloc` does.

Reply:
Ah ok - yes, malloc can allocate upper range if other address space is exhausted.

How about changing the comment to read this:
```
// On s390x it is theoretically possible to have high bit set but in practice the swift/system allocators do not allocate in this range.
```

Reply:
So it'll happen but only if you allocate ~9.2 exabytes of memory?